wasm/src/language_service.rs

```diff
@@ -1,40 +1,194 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT License.
 
+use std::{future::Future, path::PathBuf, pin::Pin, sync::Arc};
+
 use crate::{diagnostic::VSDiagnostic, serializable_type};
 use js_sys::JsString;
 use qsc::{self};
+use qsc_project::{EntryType, Manifest, ManifestDescriptor};
+use qsls::JSFileEntry;
 use rustc_hash::FxHashMap;
 use serde::{Deserialize, Serialize};
 use wasm_bindgen::prelude::*;
+use wasm_bindgen_futures::JsFuture;
 
 #[wasm_bindgen]
 pub struct LanguageService(qsls::LanguageService<'static>);
 
+/// This macro calls an async JS function, awaits it, and then applies a transformer function to it. Ultimately, it returns a function that accepts a String
+/// and returns a JS promise that is represented by a Rust future. (I know. Ouch. My head.)
+macro_rules! call_async_js_fn {
+    ($js_async_fn: ident, $transformer: expr) => {{
+        let $js_async_fn = to_js_function($js_async_fn.obj, stringify!($js_async_fn));
+
+        let $js_async_fn = move |input: String| {
+            let path = JsValue::from_str(&input);
+            let res: js_sys::Promise = $js_async_fn
+                .call1(&JsValue::NULL, &path)
+                .expect("callback should succeed")
+                .into();
+
+            let res: JsFuture = res.into();
+
+            let input = input.clone();
```

Comment:
Why clone the input here? This behavior feels very specific to `read_file`, since `read_file` basically returns its input argument as output (forgot why), but for a generic wrapper macro, it seems arbitrary.

If you feel that the use of this macro will be limited to this file, it's probably ok. If we want to make it a shared utility I'd prefer to clean it up a bit.

Reply:
Oh also I believe you're calling `.clone()` twice? What happens if you remove the `.clone()` here and let the closure below own `input`?